### PR TITLE
HTTP2: Close encoder and decoder on channelInactive and initialize clientPrefaceString on handlerAdded.

### PR DESCRIPTION
**Pull request by buchgr**
Motivation:

- The encoder and decoder should be closed right after the handler releases its resources.
- The clientPrefaceString is allocated in the constructor but releases in handlerRemoved.
  If the handler is never added to the pipeline, the clientPrefaceString will never be
  released.

Modifications:

- Call encoder.close() and decoder.close() on channelInactive.
- Release the clientPrefaceString on handlerRemoved.

Result:

- The encoder and decoder get closed right after the handler's resources are freed.
- It's easier to verify that the clientPrefaceString will also get released.

### CODE REVIEW

**Comment by buchgr** [comment]
@nmittler @Scottmitch 

I mainly care about that the encoder and decoder are closed when the channel gets inactive, but I thought it makes sense to also simplify the logic on when the clientPrefaceString gets allocated / deallocated.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 1]
```diff
@@ -156,14 +154,22 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
...
     }
 
+    /**
+     * Releases the {@code clientPrefaceString}. Any active streams will be left in the
+     * open.
+     */
     @Override
     protected void handlerRemoved0(ChannelHandlerContext ctx) throws Exception {
-        dispose();
+        if (clientPrefaceString != null) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Are we guaranteed that this is called in all cases that lead to the channel shutting down?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 2]
```diff
@@ -192,6 +198,10 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
         for (Http2Stream s : streams.toArray(new Http2Stream[streams.size()])) {
             closeStream(s, future);
         }
+
+        encoder().close();
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Should we make sure that both are closed? I.e. add exception handling?

**Comment by nmittler** [comment]
@buchgr looks good ... just a couple questions

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 2 above)
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @nmittler good point. will do.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 1 above)
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @nmittler to the best of my knowledge yes. it's now only allocated when the handler is added to the pipeline and released if 1) the handler is removed from the pipeline before the channel is closed or 2) the channel gets closed (either by closed or inactive) which will then trigger handler removed.

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @normanmaurer any insight here? Will `handlerRemoved0` always get called for a given handler?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 3]
```diff
@@ -156,14 +154,22 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
...
     public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        clientPrefaceString = clientPrefaceString(encoder.connection());
         // This handler was just added to the context. In case it was handled after
         // the connection became active, send the connection preface now.
         sendPreface(ctx);
     }
 
+    /**
+     * Releases the {@code clientPrefaceString}. Any active streams will be left in the
+     * open.
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: nit: can this fit on 1 line?

**Comment by Scottmitch** [comment]
@buchgr - LGTM just 1 nit.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 3 above)
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @Scottmitch it won't be easy, but I ll give it a try :-))

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 4]
```diff
@@ -187,10 +193,18 @@ public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exce
...
+            final Collection<Http2Stream> streams = connection().activeStreams();
+            for (Http2Stream s : streams.toArray(new Http2Stream[streams.size()])) {
+                closeStream(s, future);
+            }
+        } finally {
+            try {
+                encoder().close();
+            } finally {
+                decoder().close();
+            }
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @nmittler So this ensures that close is called on both the encoder and decoder. If exceptions are thrown in multiple places, on the last exception will be propagted. The others are "swallowed".

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: understandably ugly, but make sense :)

**Comment by nmittler** [comment]
@buchgr is everything ready to cherry-pick?

**Comment by buchgr** [comment]
@nmittler yes! thanks.

**Comment by nmittler** [comment]
cherry-picked into 4.1 and master.

**Pull request closed without merging**